### PR TITLE
Implement Expanded to Compact Difficulty Conversion

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We need to check if the compact difficulty in the block header satisfies the difficulty adjustment contextual validation rules.

Those rules require us to perform calculations on `U256` expanded difficulty values, and convert the results to `CompactDifficulty`.

This PR covers the first stage of these checks: converting expanded to compact difficulty.
(And adds a bunch of tests and cleanups.)

The next step is to design contextual validation for difficulty, see ticket #1036.

## Solution

* convert expanded to compact difficulty according to the Zcash specification
* add specific unit tests for difficulty edge cases
* use the block test vectors to in the difficulty unit tests
* add proptests for expanded to compact conversions and round-trips
* add missing proptests for the other difficulty types
* do other minor cleanups

## Related Issues

Difficulty Tracking #802 
Difficulty Contextual Validation Design #1036

## Other Resources

zcashd implementation:
* https://github.com/zcash/zcash/blob/master/src/arith_uint256.cpp#L225
* https://github.com/zcash/zcash/blob/master/src/pow.cpp#L113

### CODE REVIEW

File: zebra-chain/src/work/arbitrary.rs [hunk 1]
```diff
@@ -17,3 +19,81 @@ impl Arbitrary for equihash::Solution {
 
     type Strategy = BoxedStrategy<Self>;
 }
+
+impl Arbitrary for CompactDifficulty {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (vec(any::<u8>(), 32))
+            .prop_filter_map("zero CompactDifficulty values are invalid", |v| {
```
Comment (yaahc) [inline zebra-chain/src/work/arbitrary.rs]: This worries me. Proptest pretty aggressively advises against using any of the `filter` variants, and I'm guessing this one is going to return `None` quite often, so this might break our proptests or make them super slow. I'm guessing there's no easy way to generate random difficulties though...

File: zebra-chain/src/work/arbitrary.rs [hunk 2]
```diff
@@ -17,3 +19,81 @@ impl Arbitrary for equihash::Solution {
...
+    fn arbitrary_with(_args: ()) -> Self::Strategy {
+        // In the Zcash protocol, a Work is converted from an ExpandedDifficulty.
+        // But some randomised difficulties are impractically large, and will
+        // never appear in any real-world block. So we just use a random Work value.
+        (any::<u128>())
+            .prop_filter_map("zero Work values are invalid", |w| {
+                if w == 0 {
+                    None
+                } else {
+                    Work(w).into()
```
Comment (yaahc) [inline zebra-chain/src/work/arbitrary.rs]: Is this `into` just wrapping the `Work` with Some? I'd prefer to be explicit on this one

File: zebra-chain/src/work/difficulty.rs [hunk 3]
```diff
@@ -74,15 +73,20 @@ impl fmt::Debug for CompactDifficulty {
...
 /// conversions to `CompactDifficulty`, even if the original
 /// `ExpandedDifficulty` values are known.
+///
+/// Callers should avoid constructing `ExpandedDifficulty` zero
+/// values, because they are rejected by the consensus rules.
+//
+// TODO: Use NonZeroU256, when available
 #[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
-pub struct ExpandedDifficulty(U256);
+pub struct ExpandedDifficulty(pub(super) U256);
```
Comment (yaahc) [inline zebra-chain/src/work/difficulty.rs]: this `pub(super)` makes me think that we might want to make a new `arbitrary` submodule for just this file rather than exposing our inner value to part of the crate graph

File: zebra-chain/src/work/tests/prop.rs [hunk 4]
```diff
@@ -129,3 +130,205 @@ fn equihash_prop_test_input() -> color_eyre::eyre::Result<()> {
...
+    fn prop_difficulty_conversion(expanded_seed in any::<block::Hash>()) {
+        let expanded_seed = ExpandedDifficulty::from_hash(&expanded_seed);
+
+        let hash_zero = block::Hash([0; 32]);
+        let hash_max = block::Hash([0xff; 32]);
+        prop_assert!(expanded_seed >= hash_zero);
+        prop_assert!(expanded_seed <= hash_max);
+
+        // Skip invalid seeds
+        prop_assume!(expanded_seed != hash_zero);
```
Comment (yaahc) [inline zebra-chain/src/work/tests/prop.rs]: If this happens I feel like its a bug in our `block::Hash` Arbitrary impl

File: zebra-chain/src/work/tests/prop.rs [hunk 5]
```diff
@@ -129,3 +130,205 @@ fn equihash_prop_test_input() -> color_eyre::eyre::Result<()> {
...
+
+        let compact = expanded_seed.to_compact();
+        let expanded_trunc = compact.to_expanded();
+        let work = compact.to_work();
+
+        let work_zero = Work(0);
+        let work_max = Work(u128::MAX);
+
+        if let Some(expanded_trunc) = expanded_trunc {
```
Comment (yaahc) [inline zebra-chain/src/work/tests/prop.rs]: I'm worried that we're not ever hitting this branch or other similar branches like the one below for work, and that some of the proptests based on `filter`s aren't actually going to trigger before they exceed the reject count. We should verify that all of these proptests are working as expected.

File: zebra-chain/src/work/arbitrary.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/work/arbitrary.rs]: This function only returns `None` when the random bytes are `[0; 32]`, which happens in uniformly distributed data with probability `1 in 2^256`. (This is much smaller than the probability of a memory or disk error.)

But proptest uses some fixed values in its tests, so `[0; 32]` could be chosen more often - but it should only be one case among thousands in the test run.

File: zebra-chain/src/work/arbitrary.rs [hunk 6]
```diff
@@ -17,3 +19,81 @@ impl Arbitrary for equihash::Solution {
...
+                bytes.copy_from_slice(v.as_slice());
+                if bytes == [0; 32] {
+                    return None;
+                }
+                // In the Zcash protocol, a CompactDifficulty is generated using the difficulty
+                // adjustment functions. Instead of using those functions, we make a random
+                // ExpandedDifficulty, then convert it to a CompactDifficulty.
+                ExpandedDifficulty::from_hash(&block::Hash(bytes))
+                    .to_compact()
+                    .into()
```
Comment (teor2345) [inline zebra-chain/src/work/arbitrary.rs]:
```suggestion
                Some(ExpandedDifficulty::from_hash(&block::Hash(bytes))
                    .to_compact())
```

File: zebra-chain/src/work/arbitrary.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/work/arbitrary.rs]: Thanks for the style tip, I'll check every `into` in this PR, and in the difficulty module.

File: zebra-chain/src/work/difficulty.rs (hunk 3 above)
Comment (teor2345) [reply zebra-chain/src/work/difficulty.rs]: I'm not quite sure how this would look, can you fill in some more of the details?
I find Rust modules confusing sometimes.

File: zebra-chain/src/work/tests/prop.rs (hunk 4 above)
Comment (teor2345) [reply zebra-chain/src/work/tests/prop.rs]: I'm not sure I understand why this is a bug.

All-zeroes is a valid hash.

File: zebra-chain/src/work/arbitrary.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/work/arbitrary.rs]: Note that `ExpandedDifficulty::to_compact` is infallible, and always returns a `CompactDifficulty`.

File: zebra-chain/src/work/tests/prop.rs (hunk 5 above)
Comment (teor2345) [reply zebra-chain/src/work/tests/prop.rs]: How do you know that there will be a large number of `work` rejections?

Here's how `work` is calculated:
```
work = 2^256 / expanded
```

So if the original `expanded` value is large (and most uniformly distributed values are large), the corresponding `work` value should be small.

Specifically, I expect an distribution of `work` values similar to:
```
1: 1/2 of all values
2: 1/2-1/3 = 1/6 of all values
3: 1/3-1/4 = 1/12 of all values
4: 1/4-1/5 = 1/20 of all values
...
```

But it's easy enough to move the `Work` checks into functions that generate `any::<Work>`. (Or move the non-`Work` checks before any `prop_assume!`s.) So I can just do that as a precaution.

**Review by teor2345 — COMMENTED** [review]
I'll make these changes over the next day or so.

File: zebra-chain/src/work/difficulty.rs (hunk 3 above)
Comment (yaahc) [reply zebra-chain/src/work/difficulty.rs]: items defined in a module are also visible to all sub modules regardless of the visibility modifier so you wouldn't need pub super if the code that accesses these methods was moved to a sub module of this module

File: zebra-chain/src/work/tests/prop.rs (hunk 4 above)
Comment (yaahc) [reply zebra-chain/src/work/tests/prop.rs]: oh my mistake never mind then

File: zebra-chain/src/work/arbitrary.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/work/arbitrary.rs]: Ping @yaahc - I'd like to finish up this PR soon, what do you think of my explanations here?

File: zebra-chain/src/work/arbitrary.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/work/arbitrary.rs]: Fixed in 50c917ce578e9feace3a805d45e149a395edd9d0

File: zebra-chain/src/work/difficulty.rs (hunk 3 above)
Comment (teor2345) [reply zebra-chain/src/work/difficulty.rs]: Fixed in 06ce32931848a090fea1a7072c823736c62c5027, including CompactDifficulty, ExpandedDifficulty, and Work.

File: zebra-chain/src/work/tests/prop.rs (hunk 5 above)
Comment (teor2345) [reply zebra-chain/src/work/tests/prop.rs]: Split Work from other proptests in 808f6317e62b8025b3c3ae427d6df2e26ccc51cd.

I don't know how to verify if proptests are working.

I tried setting `PROPTEST_MAX_LOCAL_REJECTS` and `PROPTEST_MAX_GLOBAL_REJECTS` to 0 and 1, and the tests still passed.

But I don't know if "abort" means "fail the test" or "skip the test".

https://docs.rs/proptest/0.10.1/proptest/test_runner/struct.Config.html#structfield.max_local_rejects

**Comment by teor2345** [comment]
I opened #1231 to follow up the Windows failure at:
https://github.com/ZcashFoundation/zebra/pull/1196/checks?check_run_id=1320727402#step:7:1397

I'll close and re-open the PR to re-run the checks.

File: zebra-chain/src/work/arbitrary.rs (hunk 1 above)
Comment (yaahc) [reply zebra-chain/src/work/arbitrary.rs]: sounds good

File: zebra-chain/src/work/tests/prop.rs (hunk 5 above)
Comment (yaahc) [reply zebra-chain/src/work/tests/prop.rs]: sounds like its good, I'd have just put some tracing events into each of the branches to ensure that they were emitting output, but I think if it passes with multiple iterations and minimal rejects allowed, which afaik counter filtered out inputs, then I think we're good.

**Review by yaahc — APPROVED** [review]
